### PR TITLE
windows-compatible runfiles

### PR DESCRIPTION
**Pull request by alexeagle**

### CODE REVIEW

**Comment by alexeagle** [comment]
Tested manually on windows:
```
$ bazel build examples/rollup
$ bazel-out/msvc_x64-fastbuild/bin/examples/rollup/rollup --help
```

/cc @dslomov

File: internal/node_launcher.sh [hunk 1]
```diff
@@ -92,4 +92,8 @@ done
 # Note: for debugging it is useful to see what files are actually present, eg:
 # find . -name thingImLookingFor
 
-exec "${RUNFILES}/TEMPLATED_node" "${NODE_OPTIONS[@]}" "${RUNFILES}/TEMPLATED_script_path" "${ARGS[@]}"
+# Lookup the real paths from the runfiles manifest.
+NODE_PATH=$(grep TEMPLATED_node ${RUNFILES} | awk '{print $2}')
```
Comment (ocombe) [inline internal/node_launcher.sh]: this doesn't seem very robust & optimized, is there an alternative?

Comment (alexeagle) [reply internal/node_launcher.sh]: It certainly isn't very portable.

/cc @mprobst for ideas

File: internal/node_loader.js [hunk 2]
```diff
@@ -56,8 +58,20 @@ function resolveToModuleRoot(path) {
   return null;
 }
 
-function runfilesDir() {
-  return process.env.RUNFILES || process.env.TEST_SRCDIR;
+function loadRunfiles(manifestPath) {
```
Comment (mprobst) [inline internal/node_loader.js]: this is loading the manifest, not the runfiles, right?

File: internal/node_loader.js [hunk 3]
```diff
@@ -56,8 +58,20 @@ function resolveToModuleRoot(path) {
...
+  return new Promise((resolve, reject) => {
+    const map = Object.create(null);
+    const input = fs.createReadStream(manifestPath, {encoding: 'utf-8'});
+    const manifest = readline.createInterface({input});
+
+    manifest.on('line', line => {
+      if (!line) return;
+      const [key, value] = line.split(" ");
```
Comment (mprobst) [inline internal/node_loader.js]: can you give better names than key/value here?

File: internal/node_loader.js [hunk 4]
```diff
@@ -56,8 +58,20 @@ function resolveToModuleRoot(path) {
...
 
-function runfilesDir() {
-  return process.env.RUNFILES || process.env.TEST_SRCDIR;
+function loadRunfiles(manifestPath) {
+  return new Promise((resolve, reject) => {
+    const map = Object.create(null);
+    const input = fs.createReadStream(manifestPath, {encoding: 'utf-8'});
+    const manifest = readline.createInterface({input});
+
+    manifest.on('line', line => {
```
Comment (mprobst) [inline internal/node_loader.js]: what happens when reading fails here?

File: internal/node_loader.js [hunk 5]
```diff
@@ -94,13 +110,17 @@ module.constructor._resolveFilename =
   throw error;
 }
 
+let runfilesManifest;
```
Comment (mprobst) [inline internal/node_loader.js]: docs?

File: internal/node_loader.js [hunk 6]
```diff
@@ -94,13 +110,17 @@ module.constructor._resolveFilename =
...
   // NB: entry_point below is replaced during the build process.
   var mainScript = process.argv[1] = 'TEMPLATED_entry_point';
   try {
-    module.constructor._load(mainScript, this, /*isMain=*/true);
+    loadRunfiles(process.env.RUNFILES).then(mf => {
+      runfilesManifest = mf;
+      module.constructor._load(mainScript, this, /*isMain=*/true);
+    });
   } catch (e) {
     console.error('failed to load main ', e.stack || e);
```
Comment (mprobst) [inline internal/node_loader.js]: now that the code above is async, won't this exit the catch block before anything happens?

File: internal/node_launcher.sh [hunk 7]
```diff
@@ -46,16 +46,16 @@ esac
 
 if [[ -n "$TEST_SRCDIR" ]]; then
   # Case 4, bazel has identified runfiles for us.
```
Comment (mprobst) [inline internal/node_launcher.sh]: the docs talk about finding our runfiles tree, but this is finding our manifest. Update?

File: internal/node_launcher.sh (hunk 1 above)
Comment (alexeagle) [reply internal/node_launcher.sh]: Updated to use only bash, and test for equality of the templated bit, so it should not need escaping

File: internal/node_launcher.sh (hunk 7 above)
Comment (alexeagle) [reply internal/node_launcher.sh]: done

File: internal/node_loader.js (hunk 4 above)
Comment (mprobst) [reply internal/node_loader.js]: Maybe this helps?

https://stackoverflow.com/questions/21771220/error-handling-with-node-js-streams

File: internal/node_launcher.sh [hunk 8]
```diff
@@ -92,4 +92,20 @@ done
...
-exec "${RUNFILES}/TEMPLATED_node" "${NODE_OPTIONS[@]}" "${RUNFILES}/TEMPLATED_script_path" "${ARGS[@]}"
+# Lookup the real paths from the runfiles manifest with no dependency on posix
+while read line; do
+  declare -a PARTS=($line)
+  if [ "${PARTS[0]}" == "TEMPLATED_node" ]; then
+    readonly node="${PARTS[1]}"
+  elif [ "${PARTS[0]}" == "TEMPLATED_script_path" ]; then
+    readonly script="${PARTS[1]}"
+  fi
+done < ${RUNFILES}
```
Comment (mprobst) [inline internal/node_launcher.sh]: quotes around runfiles?

**Comment by mprobst** [comment]
You haven't resolved a bunch of comments from me, and also haven't waited for an LGTM. Was that intentional?

File: internal/node_loader.js (hunk 4 above)
Comment (alexeagle) [reply internal/node_loader.js]: I don't think that applies to the `manifest` object here, the return of `readline.createInterface` is not a stream. But I think I could listen for `error` events on the underlying stream?
It really needs unit tests, I don't know how to verify the error handling. Maybe it's time for `jasmine_node_test`

**Pull request merged**